File: src/cursor_test.zig
const std = @import("std");
const Blake3 = std.crypto.hash.Blake3;
const expect = std.testing.expect;
const expectEqual = std.testing.expectEqual;
const expectEqualSlices = std.testing.expectEqualSlices;

const lmdb = @import("lmdb");

const K = 32;
const Q = 4;
const Header = @import("cursor.zig").Header(K, Q);
const Builder = @import("builder.zig").Builder(K, Q);
const Cursor = @import("cursor.zig").Cursor(K, Q);
const keys = @import("keys.zig");
const utils = @import("utils.zig");

const allocator = std.heap.c_allocator;

test "basic cursor operations" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    const env = try utils.open(tmp.dir, .{});
    defer env.deinit();

    const txn = try env.transaction(.{ .mode = .ReadWrite });
    defer txn.abort();

    const db = try txn.database(null, .{});

    var builder = try Builder.init(allocator, db, .{});
    defer builder.deinit();

    try builder.set("a", "foo");
    try builder.set("b", "bar");
    try builder.set("c", "baz");
    try builder.build();

    var cursor = try Cursor.init(allocator, db);
    defer cursor.deinit();

    const root = try cursor.goToRoot();
    try expect(root.level == 3);
    try expect(root.key == null);
    try expect(root.value == null);

    if (try cursor.seek(0, "a")) |node| {
        try expectEqual(0, node.level);
        try keys.expectEqual("a", node.key);
        try keys.expectEqual("foo", node.value);
    } else return error.NotFound;

    if (try cursor.goToNext()) |node| {
        try expectEqual(0, node.level);
        try keys.expectEqual("b", node.key);
        try keys.expectEqual("bar", node.value);
    } else return error.NotFound;

    if (try cursor.goToNext()) |node| {
        try expectEqual(0, node.level);
        try keys.expectEqual("c", node.key);
        try keys.expectEqual("baz", node.value);
    } else return error.NotFound;
}

File: test/cursor.test.js
import path from "node:path";

import test from "ava";

import * as okra from "../index.js";

import { tmpdir } from "./utils.js";

test(
  "Open cursor",
  tmpdir((t, directory) => {
    const tree = new okra.Tree(path.resolve(directory, "data.okra"));

    {
      const txn = new okra.Transaction(tree, { readOnly: false });
      txn.set(Buffer.from("a"), Buffer.from("foo"));
      txn.set(Buffer.from("b"), Buffer.from("baz"));
      txn.set(Buffer.from("c"), Buffer.from("bar"));
      txn.commit();
    }

    {
      const txn = new okra.Transaction(tree, { readOnly: true });
      const cursor = new okra.Cursor(txn);
      const root = cursor.goToRoot();
      t.deepEqual(root, {
        level: 1,
        key: null,
        hash: Buffer.from("34c52f171a19da9c861ec3f469c675d6", "hex"),
      });

      cursor.close();
      txn.abort();
    }

    tree.close();
  }),
);
